src/components/shapes/attributes.js

```diff
@@ -7,19 +7,80 @@ var dash = require('../drawing/attributes').dash;
 var extendFlat = require('../../lib/extend').extendFlat;
 var templatedArray = require('../../plot_api/plot_template').templatedArray;
 var axisPlaceableObjs = require('../../constants/axis_placeable_objects');
+var basePlotAttributes = require('../../plots/attributes');
 var shapeTexttemplateAttrs = require('../../plots/template_attributes').shapeTexttemplateAttrs;
 var shapeLabelTexttemplateVars = require('./label_texttemplate');
 
 module.exports = templatedArray('shape', {
-    visible: {
+    visible: extendFlat({}, basePlotAttributes.visible, {
+        editType: 'calc+arraydraw',
+        description: [
+            'Determines whether or not this shape is visible.',
+            'If *legendonly*, the shape is not drawn,',
+            'but can appear as a legend item',
+            '(provided that the legend itself is visible).'
+        ].join(' ')
+    }),
+
+    showlegend: {
         valType: 'boolean',
-        dflt: true,
+        dflt: false,
         editType: 'calc+arraydraw',
         description: [
-            'Determines whether or not this shape is visible.'
+            'Determines whether or not this',
+            'shape is shown in the legend.'
         ].join(' ')
     },
 
+    legend: extendFlat({}, basePlotAttributes.legend, {
+        editType: 'calc+arraydraw',
+        description: [
+            'Sets the reference to a legend to show this shape in.',
+            'References to these legends are *legend*, *legend2*, *legend3*, etc.',
+            'Settings for these legends are set in the layout, under',
+            '`layout.legend`, `layout.legend2`, etc.'
+        ].join(' ')
+    }),
+
+    legendgroup: extendFlat({}, basePlotAttributes.legendgroup, {
+        editType: 'calc+arraydraw',
+        description: [
+            'Sets the legend group for this shape.',
+            'Traces and shapes part of the same legend group hide/show at the same time',
+            'when toggling legend items.'
+        ].join(' ')
+    }),
+
+    legendgrouptitle: {
+        text: extendFlat({}, basePlotAttributes.legendgrouptitle.text, {
+            editType: 'calc+arraydraw'
+        }),
+        font: fontAttrs({
+            editType: 'calc+arraydraw',
+            description: [
+                'Sets this legend group\'s title font.'
+            ].join(' '),
+        }),
+        editType: 'calc+arraydraw',
+    },
+
+    legendrank: extendFlat({}, basePlotAttributes.legendrank, {
+        editType: 'calc+arraydraw',
+        description: [
+            'Sets the legend rank for this shape.',
+            'Items and groups with smaller ranks are presented on top/left side while',
+            'with *reversed* `legend.traceorder` they are on bottom/right side.',
+            'The default legendrank is 1000,',
+            'so that you can use ranks less than 1000 to place certain items before all unranked items,',
+            'and ranks greater than 1000 to go after all unranked items.'
```

Comment:
Can we say something about unranked or equal rank items? I presume traces according to their order in `data`, then shapes according to their order in `layout.shapes`?

Also I don't see `shapes[].legendrank` used in a test image to intersperse shapes and traces, or `shapes[].legend` to put shapes on multiple legends, can we add those?

Reply:
Good call.
Addressed in [2e77fc7](https://github.com/plotly/plotly.js/pull/6653/commits/2e77fc724093317c93ce234abb961bda87929b4f), [58b1cb1](https://github.com/plotly/plotly.js/pull/6653/commits/58b1cb1f7afa921efefbc656686f9677e2b22ebf) and [3168078](https://github.com/plotly/plotly.js/pull/6653/commits/316807817b02d8339c3ba498123fce5f574e19aa).